pkg/controller/elasticsearch/remotecluster/elasticsearch.go

```diff
@@ -25,88 +26,143 @@ var log = logf.Log.WithName("remotecluster")
 		}
 	}
 
-	// RemoteClusters to remove
-	for name := range currentRemoteClusters {
-		if _, ok := expectedRemoteClusters[name]; !ok {
-			log.Info("Removing remote cluster",
-				"namespace", es.Namespace,
-				"es_name", es.Name,
-				"remote_cluster", name,
-			)
-			remoteClusters[name] = esclient.RemoteCluster{Seeds: nil}
-		}
+	remoteClustersToUpdate := make([]string, 0, len(remoteClustersInSpec)) // only used for logging
+	// remoteClustersToApply are clusters to add (or update) based on what is specified in the Elasticsearch spec.
+	remoteClustersToApply := make(map[string]esclient.RemoteCluster)
+	for name, remoteCluster := range remoteClustersInSpec {
+		remoteClustersToUpdate = append(remoteClustersToUpdate, name)
+		// Declare remote cluster in ES
+		seedHosts := []string{services.ExternalTransportServiceHost(remoteCluster.ElasticsearchRef.NamespacedName())}
+		remoteClustersToApply[name] = esclient.RemoteCluster{Seeds: seedHosts}
+		// Ensure this cluster is tracked in the annotation
+		remoteClustersInAnnotation[name] = struct{}{}
 	}
 
-	if len(remoteClusters) > 0 {
+	// RemoteClusters to remove from Elasticsearch
+	for _, name := range remoteClustersToDelete {
+		remoteClustersToApply[name] = esclient.RemoteCluster{Seeds: nil}
+	}
+
+	// Update the annotation
+	if err := annotateWithCreatedRemoteClusters(c, es, remoteClustersInAnnotation); err != nil {
+		return true, err
+	}
+
+	// Since the annotation is update before Elasticsearch we should requeue to sync the annotation
+	// if some clusters are deleted from Elasticsearch.
+	shouldRequeue := len(remoteClustersToDelete) > 0
+	if len(remoteClustersToApply) > 0 {
 		// Apply the settings
-		if err := updateSettings(esClient, remoteClusters); err != nil {
-			return err
-		}
-		// Update the annotation
-		return annotateWithRemoteClusters(c, es, expectedRemoteClusters)
+		sort.Strings(remoteClustersToUpdate)
+		sort.Strings(remoteClustersToDelete)
+		log.Info("Updating remote cluster settings",
+			"namespace", es.Namespace,
+			"es_name", es.Name,
+			"updated_remote_clusters", remoteClustersToUpdate,
+			"deleted_remote_clusters", remoteClustersToDelete,
+		)
+		return shouldRequeue, updateSettings(esClient, remoteClustersToApply)
 	}
-	return nil
+	return shouldRequeue, nil
 }
 
-// getExpectedRemoteClusters returns a map with the expected remote clusters
+// getRemoteClustersInElasticsearch returns all the remote clusters currently declared in Elasticsearch
+func getRemoteClustersInElasticsearch(esClient esclient.Client) (map[string]struct{}, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), esclient.DefaultReqTimeout)
+	defer cancel()
+	remoteClustersInEs := make(map[string]struct{})
+	remoteClusterSettings, err := esClient.GetRemoteClusterSettings(ctx)
+	if err != nil {
+		return remoteClustersInEs, err
+	}
+	for remoteClusterName := range remoteClusterSettings.PersistentSettings.Cluster.RemoteClusters {
```

Comment:
I guess we can ignore transient settings for the purposes of this function?

Reply:
This is the assumption I made since we are only interested in clusters created by ECK.